apps/rush/src/RushVersionSelector.ts

```diff
@@ -27,27 +30,45 @@ export class RushVersionSelector {
       { node: process.versions.node }
     );
 
+    let installPromise: Promise<void> = Promise.resolve();
+
     if (!installMarker.isValid()) {
-      // Need to install Rush
-      console.log(`Rush version ${version} is not currently installed. Installing...`);
-
-      Utilities.installPackageInDirectory(
-        expectedRushPath,
-        isLegacyRushVersion ? '@microsoft/rush' : '@microsoft/rush-lib',
-        version,
-        'rush-local-install',
-        MAX_INSTALL_ATTEMPTS,
-        true
-      );
-
-      console.log(`Successfully installed Rush version ${version} in ${expectedRushPath}`);
-    }
+      installPromise = installPromise.then(() => {
+        // Need to install Rush
+        console.log(`Rush version ${version} is not currently installed. Installing...`);
+
+        const resourceName: string = `rush-${version}`;
+
+        console.log(`Trying to acquire lock for ${resourceName}`);
 
-    // If we've made it here without exception, write the flag file
-    installMarker.create();
+        return LockFile.acquire(expectedRushPath, resourceName)
+          .then((lock: LockFile) => {
 
-    if (semver.lt(version, '3.0.20')) {
-      return () => {
+            if (installMarker.isValid()) {
+              console.log('Another process performed the installation.');
+            } else {
+              Utilities.installPackageInDirectory(
+                expectedRushPath,
+                isLegacyRushVersion ? '@microsoft/rush' : '@microsoft/rush-lib',
+                version,
+                'rush-local-install',
+                MAX_INSTALL_ATTEMPTS,
+                true
+              );
+
+              console.log(`Successfully installed Rush version ${version} in ${expectedRushPath}.`);
+
+              // If we've made it here without exception, write the flag file
+              installMarker.create();
+
+              lock.release();
+            }
```

Comment:
Two rush processes are invoked at the same time, will the one that runs second reinstall rush-lib after the first one finishes installing it?

Reply:
No, it will print "Another process performed the installation." from line 48.  I validated this during testing.

---
In reply to: [185168976](https://github.com/Microsoft/web-build-tools/pull/632#discussion_r185168976) [](ancestors = 185168976)